File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.ts
/**
 * Definition for _Node.
 * class _Node {
 *     val: number
 *     next: _Node | null
 *     random: _Node | null
 * 
 *     constructor(val?: number, next?: _Node, random?: _Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */


function copyRandomList(head: _Node | null): _Node | null {
    let temp = head;
    let mp = new Map<_Node | null , _Node | null>();

    while(temp !== null) {
        let newNode = new _Node(temp.val);
        mp.set(temp , newNode);
        temp = temp.next;
    }
    console.log(mp);

    temp = head;
    while(temp !== null) {
        let copyNode = mp.get(temp) || null;
        copyNode.next = mp.get(temp.next) || null;
        copyNode.random = mp.get(temp.random);
        temp = temp.next;
    }

    return mp.get(head);
};

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Map<Node , Node> mp = new HashMap<>();
        Node curNode = head;
        while(curNode != null) {
            Node newNode = new Node(curNode.val);
            mp.put(curNode , newNode);
            curNode = curNode.next;
        }

        curNode = head;
        while(curNode != null) {
            Node newNextNode = mp.get(curNode.next);
            Node randomNextNode = mp.get(curNode.random);
            Node nextNode = mp.get(curNode);
            nextNode.next = newNextNode;
            nextNode.random = randomNextNode;
            curNode = curNode.next;
        }

        return mp.get(head);
    }
}